### PR TITLE
Push notification implementation

### PR DESCRIPTION
**Pull request by cammellos**
Implements push notification server & client.

The way it is implemented is that any client can just act as a push notification server, so you can run this on your android phone if you wish, as it's just a waku/whisper client.

I have heavily commented the code as the functioning is rather complex, the server is pretty straightforward, while the client is quite complicated.

Is not quite finished, there's still some things to handle and improved, but it's at a stage that is ready to be integrated.

There are full e2e tests of the functionalities  in `protocol/push_notification_test.go`. 

I have left the commit split and committed often to ease the review process, it's quite a lot of code unfortunately.

I haven't quite updated the specs with the latest changes, I will do that tomorrow, but most of it is there: https://github.com/status-im/specs/blob/master/docs/raw/push-notification-server.md

UPDATE: specs PR https://github.com/status-im/specs/pull/146

Things that have also changed:
1) Is possible to send messages with ephemeral keys now
2) We can install multiple identity keys and receive messages from it
3) We can optionally skip datasync for some messages 
4) Moved some shared code under `common`

### CODE REVIEW

**Comment by cammellos** [comment]
@Samyoul @jakubgs @andremedeiros I have updated the specs to the latest changes https://github.com/status-im/specs/pull/146

File: multiaccounts/accounts/database.go [hunk 1]
```diff
@@ -59,25 +59,34 @@ type Settings struct {
...
-	WakuBloomFilterMode    bool             `json:"waku-bloom-filter-mode,omitempty"`
+	// NotificationsEnabled indicates whether local notifications should be enabled (android only)
+	NotificationsEnabled bool             `json:"notifications-enabled?,omitempty"`
+	PhotoPath            string           `json:"photo-path"`
+	PinnedMailserver     *json.RawMessage `json:"pinned-mailservers,omitempty"`
+	PreferredName        *string          `json:"preferred-name,omitempty"`
+	PreviewPrivacy       bool             `json:"preview-privacy?"`
+	PublicKey            string           `json:"public-key"`
+	// PushNotificationServerEnabled indicates whether we should be running a push notification server
+	PushNotificationsServerEnabled bool `json:"push-notifications-server-enabled,omitempty"`
```
Comment (andremedeiros) [inline multiaccounts/accounts/database.go]: We're not being consistent on how we name boolean fields. See above with `notifications-enabled?`. Can we decide on one and stay consistent?

File: protocol/common/message_processor.go [hunk 2]
```diff
@@ -64,50 +82,66 @@ func newMessageProcessor(
...
-func (p *messageProcessor) SendPrivate(
+func (p *MessageProcessor) SendPrivate(
 	ctx context.Context,
 	recipient *ecdsa.PublicKey,
 	rawMessage *RawMessage,
 ) ([]byte, error) {
 	p.logger.Debug(
 		"sending a private message",
-		zap.Binary("public-key", crypto.FromECDSAPub(recipient)),
+		zap.String("public-key", types.EncodeHex(crypto.FromECDSAPub(recipient))),
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:124]: 👏

File: protocol/common/message_processor.go [hunk 3]
```diff
@@ -64,50 +82,66 @@ func newMessageProcessor(
...
 		zap.String("site", "SendPrivate"),
 	)
+	// Currently we don't support sending through datasync and setting custom waku fields,
+	// as the datasync interface is not rich enough to propagate that information, so we
+	// would have to add some complexity to handle this.
+	if rawMessage.ResendAutomatically && (rawMessage.Sender != nil || rawMessage.SkipEncryption) {
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:130]: Should we check for `p.features.Datasync` here too?

File: protocol/common/message_processor.go [hunk 4]
```diff
@@ -134,51 +174,67 @@ func (p *messageProcessor) SendGroup(
...
 
 // sendPrivate sends data to the recipient identifying with a given public key.
-func (p *messageProcessor) sendPrivate(
+func (p *MessageProcessor) sendPrivate(
 	ctx context.Context,
 	recipient *ecdsa.PublicKey,
 	rawMessage *RawMessage,
 ) ([]byte, error) {
-	p.logger.Debug("sending private message", zap.Binary("recipient", crypto.FromECDSAPub(recipient)))
+	p.logger.Debug("sending private message", zap.String("recipient", types.EncodeHex(crypto.FromECDSAPub(recipient))))
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:182]: 👏

File: protocol/common/message_processor.go [hunk 5]
```diff
@@ -296,12 +361,32 @@ func (p *messageProcessor) handleMessages(shhMessage *types.Message, application
...
 
-	err := message.HandleEncryption(p.identity, publicKey, p.protocol)
-	if err == encryption.ErrDeviceNotFound {
+	// if it's an ephemeral key, we don't negotiate a topic
+	decryptionKey, skipNegotiation := p.fetchDecryptionKey(message.Dst)
+
+	err := message.HandleEncryption(decryptionKey, publicKey, p.protocol, skipNegotiation)
+
+	// if it's an ephemeral key, we don't have to handle a device not found error
+	if err == encryption.ErrDeviceNotFound && !skipNegotiation {
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:389]: I'm having a hard time seeing where this error comes from. In fact, this method only returns wrapped errors, which leads me to believe that the left condition would never evaluate to true.

File: protocol/common/message_processor.go [hunk 6]
```diff
@@ -340,15 +427,15 @@ func (p *messageProcessor) handleErrDeviceNotFound(ctx context.Context, publicKe
 	return nil
 }
 
-func (p *messageProcessor) wrapMessageV1(rawMessage *RawMessage) ([]byte, error) {
-	wrappedMessage, err := v1protocol.WrapMessageV1(rawMessage.Payload, rawMessage.MessageType, p.identity)
+func (p *MessageProcessor) wrapMessageV1(rawMessage *RawMessage) ([]byte, error) {
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:430]: Have you considered moving this method to be owned by `*RawMessage`, now that we don't need to pass the identity from `*MessageProcessor`?

File: protocol/common/message_processor.go [hunk 7]
```diff
@@ -414,10 +519,70 @@ func (p *messageProcessor) sendMessageSpec(ctx context.Context, publicKey *ecdsa
...
+	c := make(chan *SentMessage, 100)
+	p.sentMessagesSubscriptions = append(p.sentMessagesSubscriptions, c)
+	return c
+}
+
+func (p *MessageProcessor) notifyOnSentMessage(sentMessage *SentMessage) {
+	// Publish on channels, drop if buffer is full
+	for _, c := range p.sentMessagesSubscriptions {
+		select {
+		case c <- sentMessage:
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:544]: My mind is blown!

File: protocol/common/message_processor.go [hunk 8]
```diff
@@ -414,10 +519,70 @@ func (p *messageProcessor) sendMessageSpec(ctx context.Context, publicKey *ecdsa
...
+		default:
+			p.logger.Warn("subscription channel full, dropping message")
```
Comment (andremedeiros) [inline protocol/common/message_processor.go]: Can we get the message to specify which channel is full?

File: protocol/common/message_processor.go [hunk 9]
```diff
@@ -414,10 +519,70 @@ func (p *messageProcessor) sendMessageSpec(ctx context.Context, publicKey *ecdsa
...
+	return p.transport.JoinPublic(chatID)
+}
+
+// AddEphemeralKey adds an ephemeral key that we will be listening to
+// note that we never removed them from now, as waku/whisper does not
+// recalculate topics on removal, so effectively there's no benefit.
+// On restart they will be gone.
+func (p *MessageProcessor) AddEphemeralKey(privateKey *ecdsa.PrivateKey) (*transport.Filter, error) {
+	p.ephemeralKeysMutex.Lock()
+	p.ephemeralKeys[types.EncodeHex(crypto.FromECDSAPub(&privateKey.PublicKey))] = privateKey
```
Comment (andremedeiros) [inline protocol/common/message_processor.go:580]: Potentially, how much could this grow? I'm nervous about Android where we have the background service running all the time.

File: protocol/common/message_processor_test.go [hunk 10]
```diff
@@ -1,4 +1,4 @@
-package protocol
+package common
```
Comment (andremedeiros) [inline protocol/common/message_processor_test.go:1]: We should have a test here that asserts a message with a `nil` sender still gets an identity.

File: protocol/migrations/sqlite/1593087212_add_mute_chat_and_raw_message_fields.up.sql [hunk 11]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE chats ADD COLUMN muted BOOLEAN DEFAULT FALSE;
```
Comment (andremedeiros) [inline protocol/migrations/sqlite/1593087212_add_mute_chat_and_raw_message_fields.up.sql:1]: Before we ship this, we have a choice to make here -- do we want to allow all chats to notify (like every other messenger) or do we want to take the Hey approach and pick who we want to be notified about? @hesterbruikman

File: protocol/protobuf/service.go [hunk 12]
```diff
@@ -4,7 +4,7 @@ import (
 	"github.com/golang/protobuf/proto"
 )
 
-//go:generate protoc --go_out=. ./chat_message.proto ./application_metadata_message.proto ./membership_update_message.proto ./command.proto ./contact.proto ./pairing.proto
+//go:generate protoc --go_out=. ./chat_message.proto ./application_metadata_message.proto ./membership_update_message.proto ./command.proto ./contact.proto ./pairing.proto ./push_notifications.proto
```
Comment (andremedeiros) [inline protocol/protobuf/service.go:7]: Lack of glob support in `go generate` is just unsettling 😢

File: protocol/pushnotificationclient/client.go [hunk 13]
```diff
@@ -0,0 +1,1287 @@
...
+
+// subscribeForScheduledMessages subscribes for messages scheduler for dispatch
+func (c *Client) subscribeForScheduledMessages() {
+	go func() {
+		c.config.Logger.Debug("subscribing for scheduled messages")
+		subscription := c.messageProcessor.SubscribeToScheduledMessages()
+		for {
+			select {
+			case m, more := <-subscription:
+				if !more {
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/client.go:521]: TIL!

File: protocol/pushnotificationclient/client.go [hunk 14]
```diff
@@ -0,0 +1,1287 @@
...
+		// The only time we are 100% certain that we can proceed is
+		// when we have non-stale info for each device, but
+		// most devices are not going to be registered, so we'd still
+		// have to wait the maximum amount of time allowed.
+		// A better way to handle this is to set a maximum timer of say
+		// 3 seconds, but act at a tick every 200ms.
+		// That way we still are able to batch multiple push notifications
+		// but we don't have to wait every time 3 seconds, which is wasteful
+		// This probably will have to be addressed before released
+		time.Sleep(3 * time.Second)
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/client.go:625]: This is fine, for now.

Is this a future improvement we should track on a feature?

File: protocol/pushnotificationclient/persistence.go [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+	}
+
+	inVector := strings.Repeat("?, ", len(installationIDs)-1) + "?"
+
+	rows, err := p.db.Query(`SELECT server_public_key, installation_id, version, access_token, retrieved_at FROM push_notification_client_info WHERE public_key = ? AND installation_id IN (`+inVector+`)`, queryArgs...) //nolint: gosec
+
+	if err != nil {
+		return nil, err
+	}
+	var infos []*PushNotificationInfo
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/persistence.go:190]:
```go
    defer rows.Close()
```

File: protocol/pushnotificationclient/persistence.go [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+
+	return infos, nil
+}
+
+func (p *Persistence) GetPushNotificationInfoByPublicKey(publicKey *ecdsa.PublicKey) ([]*PushNotificationInfo, error) {
+	rows, err := p.db.Query(`SELECT server_public_key, installation_id, access_token, retrieved_at FROM push_notification_client_info WHERE public_key = ?`, crypto.CompressPubkey(publicKey))
+	if err != nil {
+		return nil, err
+	}
+	var infos []*PushNotificationInfo
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/persistence.go:218]:
```go
    defer rows.Close()
```

File: protocol/pushnotificationclient/persistence.go [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+func (p *Persistence) GetRetriablePushNotifications() ([]*SentNotification, error) {
+	var notifications []*SentNotification
+	rows, err := p.db.Query(`SELECT retry_count, last_tried_at, error, success, public_key, installation_id, message_id FROM push_notification_client_sent_notifications WHERE NOT success AND error = ?`, protobuf.PushNotificationReport_WRONG_TOKEN)
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/persistence.go:318]:
```go
    defer rows.Close()
```

File: protocol/pushnotificationclient/persistence.go [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+
+}
+
+func (p *Persistence) GetServers() ([]*PushNotificationServer, error) {
+	rows, err := p.db.Query(`SELECT public_key, registered, registered_at,access_token,last_retried_at, retry_count FROM push_notification_client_servers`)
+	if err != nil {
+		return nil, err
+	}
+	var servers []*PushNotificationServer
+	for rows.Next() {
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/persistence.go:349]:
```go
    defer rows.Close()
```

File: protocol/pushnotificationclient/persistence.go [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+	for _, key := range keys {
+		keyArgs = append(keyArgs, crypto.CompressPubkey(key))
+	}
+
+	inVector := strings.Repeat("?, ", len(keys)-1) + "?"
+	rows, err := p.db.Query(`SELECT public_key, registered, registered_at,access_token FROM push_notification_client_servers WHERE public_key IN (`+inVector+")", keyArgs...) //nolint: gosec
+	if err != nil {
+		return nil, err
+	}
+	var servers []*PushNotificationServer
```
Comment (andremedeiros) [inline protocol/pushnotificationclient/persistence.go:380]:
```go
    defer rows.Close()
```

File: protocol/pushnotificationserver/gorush.go [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+	case protobuf.PushNotificationRegistration_FIREBASE_TOKEN:
+		return 2
+	}
+	return 0
+}
+
+func PushNotificationRegistrationToGoRushRequest(requestAndRegistrations []*RequestAndRegistration) *GoRushRequest {
+	goRushRequests := &GoRushRequest{}
+	for _, requestAndRegistration := range requestAndRegistrations {
+
```
Comment (andremedeiros) [inline protocol/pushnotificationserver/gorush.go]: newline--

File: protocol/pushnotificationserver/persistence_test.go [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/status-im/status-go/eth-node/crypto"
+	"github.com/status-im/status-go/protocol/common"
+	"github.com/status-im/status-go/protocol/protobuf"
+	"github.com/status-im/status-go/protocol/sqlite"
+)
+
+//tmpFile, err := ioutil.TempFile("", "")
```
Comment (andremedeiros) [inline protocol/pushnotificationserver/persistence_test.go]: Eliminate dead code

**Review by andremedeiros — APPROVED** [review]
Amazing job! 👏 

I have a couple of comments around minor things, but super happy to see this shipping!

File: multiaccounts/accounts/database.go (hunk 1 above)
Comment (cammellos) [reply multiaccounts/accounts/database.go]: I think we should rename everything to camelCase, as now  there's also desktop and these are all clojurisms, but I will make it consistent for now

File: protocol/migrations/sqlite/1593087212_add_mute_chat_and_raw_message_fields.up.sql (hunk 11 above)
Comment (cammellos) [reply protocol/migrations/sqlite/1593087212_add_mute_chat_and_raw_message_fields.up.sql:1]: I think if notifications are enabled, we will notify by default on 1-to-1 and group chats. cc @hesterbruikman

**Comment by cammellos** [comment]
I will be moving to work on emojii reactions, I will merge this and address the feedback in a separate PR.

**Pull request merged**